engine/apps/slack/scenarios/distribute_alerts.py

```diff
@@ -50,68 +46,92 @@
+        alert_receive_channel = alert_group.channel
+        slack_message_already_sent = alert_group.slack_message_sent
+        should_skip_escalation_in_slack = alert_group.skip_escalation_in_slack
+
+        def _mark_alert_group_slack_message_as_sent(reason_to_skip_escalation=None):
+            if not slack_message_already_sent:
+                alert_group.slack_message_sent = True
+
+                if reason_to_skip_escalation:
+                    alert_group.reason_to_skip_escalation = reason_to_skip_escalation
+
+                alert_group.save(update_fields=["slack_message_sent", "reason_to_skip_escalation"])
+
         # do not try to post alert group message to slack if its channel is rate limited
-        if alert.group.channel.is_rate_limited_in_slack:
+        if alert_receive_channel.is_rate_limited_in_slack:
             logger.info("Skip posting or updating alert_group in Slack due to rate limit")
-            AlertGroup.objects.filter(
-                pk=alert.group.pk,
-                slack_message_sent=False,
-            ).update(slack_message_sent=True, reason_to_skip_escalation=AlertGroup.RATE_LIMITED)
+            _mark_alert_group_slack_message_as_sent(reason_to_skip_escalation=AlertGroup.RATE_LIMITED)
             return
 
-        num_updated_rows = AlertGroup.objects.filter(pk=alert.group.pk, slack_message_sent=False).update(
-            slack_message_sent=True
-        )
+        if not slack_message_already_sent:
+            # this will be the case in the event that we haven't yet created a Slack message for this alert group
+
+            slack_channel = (
+                alert_group.channel_filter.slack_channel
+                if alert_group.channel_filter
+                # if channel filter is deleted mid escalation, use default Slack channel
+                else alert_receive_channel.organization.default_slack_channel
+            )
 
-        if num_updated_rows == 1:
             try:
-                slack_channel = (
-                    alert.group.channel_filter.slack_channel
-                    if alert.group.channel_filter
-                    # if channel filter is deleted mid escalation, use default Slack channel
-                    else alert.group.channel.organization.default_slack_channel
+                self._post_alert_group_to_slack(
+                    slack_team_identity=self.slack_team_identity,
+                    alert_group=alert_group,
+                    alert=alert,
+                    attachments=alert_group.render_slack_attachments(),
+                    slack_channel=slack_channel,
+                    blocks=alert_group.render_slack_blocks(),
                 )
-                self._send_first_alert(alert, slack_channel)
+
+                _mark_alert_group_slack_message_as_sent()
             except (SlackAPIError, TimeoutError):
-                AlertGroup.objects.filter(pk=alert.group.pk).update(slack_message_sent=False)
```

Comment:
by not doing the `.update(slack_message_sent=True)` call until _after_ we successfully call `self._post_alert_group_to_slack`, I don't think we need this `.update(slack_message_sent=False)` call here.

if `self._post_alert_group_to_slack` throws an exception, `_mark_alert_group_slack_message_as_sent` will never get called, and `slack_message_sent` will never get set to True.

Reply:
I'm not sure about this part, but I assume it was here to mark ag sent asap to prevent double posting to slack in case ag got 2 alerts at the same time. Does it make sense? 🤔

Reply:
ahhh okay, I see what you mean. In that case, the second alert would result in `False` here, and we'd not double send:

```python
if not slack_message_already_sent:
```

Let me tweak this a bit/see if I can add some tests for this case. Thanks for calling this out.